x-pack/plugins/fleet/server/services/metrics/fleet_metrics_task.ts

```diff
@@ -0,0 +1,184 @@
+  ) {
+    taskManager.registerTaskDefinitions({
+      [TYPE]: {
+        title: TITLE,
+        timeout: TIMEOUT,
+        maxAttempts: 1,
+        createTaskRunner: ({ taskInstance }: { taskInstance: ConcreteTaskInstance }) => {
+          return {
+            run: async () => {
+              return withSpan({ name: TYPE, type: 'metrics' }, () =>
+                this.runTask(taskInstance, () => fetchAgentMetrics(this.abortController))
+              );
+            },
+
+            cancel: async () => {
+              this.abortController.abort('task timed out');
+            },
+          };
+        },
+      },
+    });
+  }
+
+  private runTask = async (
+    taskInstance: ConcreteTaskInstance,
+    fetchAgentMetrics: () => Promise<AgentMetrics | undefined>
+  ) => {
+    if (!this.wasStarted) {
+      appContextService.getLogger().debug('[runTask()] Aborted. Task not started yet');
+      return;
+    }
+    // Check that this task is current
+    if (taskInstance.id !== this.taskId) {
+      throwUnrecoverableError(new Error('Outdated task version for task: ' + taskInstance.id));
+      return;
+    }
+    if (!this.esClient) {
+      appContextService.getLogger().info('esClient not set, skipping Fleet metrics task');
+      return;
+    }
+    appContextService.getLogger().info('Running Fleet metrics task');
+
+    try {
+      const agentMetrics = await fetchAgentMetrics();
+      if (!agentMetrics) {
+        return;
+      }
+      const { agents_per_version: agentsPerVersion, agents } = agentMetrics;
+      const clusterInfo = await this.esClient.info();
+      const getCommonFields = (dataset: string) => {
+        return {
+          data_stream: {
+            dataset,
+            type: 'metrics',
+            namespace: 'default',
+          },
+          service: {
+            id: appContextService.getKibanaInstanceId(),
+            version: appContextService.getKibanaVersion(),
+          },
+          cluster: {
+            id: clusterInfo?.cluster_uuid ?? '',
+          },
+        };
+      };
+      const agentStatusDoc = {
+        '@timestamp': new Date().toISOString(),
+        ...getCommonFields('fleet_server.agent_status'),
+        fleet: {
+          agents: {
+            total: agents.total_all_statuses,
+            enrolled: agents.total_enrolled,
+            unenrolled: agents.unenrolled,
+            healthy: agents.healthy,
+            offline: agents.offline,
+            updating: agents.updating,
+            unhealthy: agents.unhealthy,
+            inactive: agents.inactive,
+            upgrading_step: agentMetrics.upgrading_step,
+          },
+        },
+      };
+      appContextService
+        .getLogger()
+        .trace('Agent status metrics: ' + JSON.stringify(agentStatusDoc));
+      await this.esClient.index({
+        index: 'metrics-fleet_server.agent_status-default',
+        body: agentStatusDoc,
+      });
+
+      for (const byVersion of agentsPerVersion) {
```

Comment:
nit: does it make sense to do a bulk insert for all version?

Reply:
changed to bulk